qutip/semidefinite.py

```diff
@@ -193,6 +193,78 @@ def dnorm_problem(dim):
     ))
 
     problem = cvxpy.Problem(objective, constraints)
-
+    print('problem defined')
     return problem, Jr, Ji, X, rho0, rho1
 
+
+def dnorm_sparse_problem(dim, J_dat=None):
+    # Start assembling constraints and variables.
+    constraints = []
+
+    # Make a complex variable for X.
+    X = complex_var(dim ** 2, dim ** 2, "X")
+
+    # Make complex variables for rho0 and rho1.
+    rho0 = complex_var(dim, dim, "rho0")
+    rho1 = complex_var(dim, dim, "rho1")
+    constraints += dens(rho0, rho1)
+
+    # Finally, add the tricky positive semidefinite constraint.
+    # Since we're using column-stacking, but Watrous used row-stacking,
+    # we need to swap the order in Rho0 and Rho1. This is not straightforward,
+    # as CVXPY requires that the constant be the first argument. To solve this,
+    # We conjugate by SWAP.
+    W = qudit_swap(dim).data.todense()
+    W = Complex(re=W.real, im=W.imag)
+    Rho0 = conj(W, kron(np.eye(dim), rho0))
+    Rho1 = conj(W, kron(np.eye(dim), rho1))
+
+    Y = cvxpy.bmat([
+        [Rho0.re, X.re,      -Rho0.im, -X.im],
+        [X.re.T, Rho1.re,    X.im.T, -Rho1.im],
+
+        [Rho0.im, X.im,      Rho0.re, X.re],
+        [-X.im.T, Rho1.im,   X.re.T, Rho1.re],
+    ])
+    constraints += [Y >> 0]
+
+    logger.debug("Using {} constraints.".format(len(constraints)))
+
+    J_val = sp.coo_matrix(sp.csr_matrix((J_dat.data, J_dat.indices,
+                                         J_dat.indptr),
+                          shape=J_dat.shape))
+
+    def adapt_sparse_params(A_val, dim):
+
+        A = cvxpy.Parameter((dim**2, dim**2))
+        side_size = dim**2
+        A_nnz = cvxpy.Parameter(A_val.nnz)
+
+        A_data = np.ones(A_nnz.size)
+        A_rows = A_val.row * side_size + A_val.col
+        A_cols = np.arange(A_nnz.size)
+
+        A_Indexer = sp.coo_matrix((A_data, (A_rows, A_cols)),
+                                  shape=(side_size**2, A_nnz.size))
+
+        A = cvxpy.reshape(A_Indexer @ A_nnz, (side_size, side_size), order='C')
+
+        return A_nnz, A
```

Comment:
Hi I am reproducing the first proposed solution to the problem https://github.com/cvxgrp/cvxpy/issues/1159#issuecomment-718925710 . I am not setting it inside mainly to avoid repeating code, as i am calling the function to set both Jr_nnz and Ji_nnz .  We are not able to directly call A directly if it is sparse  as cvxpy is not able to , you can only do it through https://github.com/cvxgrp/cvxpy/issues/1159#issuecomment-718170862 . 

You might use the second proposed solution from here  https://github.com/cvxgrp/cvxpy/issues/1159#issuecomment-718925710  but you get a larger number of non-zero entries as shown there.

Reply:
But the values of `A_nnz` are then set later using what's passed to this function as `A_val`.  So it'd still be encapsulated within your function, and you'd be able to use the function call for both still.  It seems odd that you sort of "leak" state by returning `A_nnz`, then you set up the optimisation problem, and only then do you mutate the `A_nnz` object using values that you actually knew long before.

Reply:
Can you leave a little comment to say approximately (~1-2 lines) what it's doing and a reference to that cvxpy issue as a code comment (something like `see GitHub issue cvxgrp/cvxpy#1159`), so in the future we'll know the reasoning?  It'll help in the future, because at some point presumably the upstream might change or get fixed, and then we'll understand what it's doing and if we can remove the workaround.